runtime/expr/function/unflatten.go

```diff
@@ -0,0 +1,153 @@
+package function
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/pkg/field"
+	"github.com/brimdata/zed/zcode"
+)
+
+// https://github.com/brimdata/zed/blob/main/docs/language/functions.md#unflatten
+type Unflatten struct {
+	builder zcode.Builder
+	stack   []record
+	zctx    *zed.Context
+
+	// These exist only to reduce memory allocations.
+	path    field.Path
+	columns []zed.Column
+}
+
+func NewUnflatten(zctx *zed.Context) *Unflatten {
+	return &Unflatten{
+		zctx: zctx,
+	}
+}
+
+func (u *Unflatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
+	val := args[0]
+	array, ok := zed.TypeUnder(val.Type).(*zed.TypeArray)
+	if !ok {
+		return &val
+	}
+	u.builder.Reset()
+	u.stack = u.stack[:0]
+	for it := val.Bytes.Iter(); !it.Done(); {
+		path, typ, vb := u.parseElem(array.Type, it.Next())
+		if typ == nil {
+			continue
+		}
+		if err := u.appendItem(0, path, typ, vb); err != nil {
+			return u.zctx.NewErrorf("unflatten: %v", err)
+		}
+	}
+	if len(u.stack) == 0 {
+		// No suitable items found in array.
+		return &val
+	}
+	if err := u.closeChildren(u.stack); err != nil {
+		return u.zctx.NewErrorf("unflatten: %v", err)
+	}
+	typ, err := u.typeOfRecord(u.stack[0])
+	if err != nil {
+		return u.zctx.NewErrorf("unflatten: %v", err)
+	}
+	return ctx.NewValue(typ, u.builder.Bytes()).Copy()
+}
+
+type record struct {
+	names []string
+	types []zed.Type
+}
+
+func (r *record) reset() {
+	r.names = r.names[0:]
+	r.types = r.types[0:]
+}
```

Comment:
You could use `[]zed.Column` instead of this new type.

Reply:
Truth.